File: 475.heaters/solution.kt
class Solution {
    fun findRadius(houses: IntArray, heaters: IntArray): Int {
        var res = 0
        
        heaters.sort()

        for (house in houses) {
            val i = heaters.binarySearch(house).let {
                if (it < 0) -(it + 1) else it
            }
            val distances = sequence {
                if (i > 0) {
                    yield(house - heaters[i - 1])
                }
                if (i < heaters.size) {
                    yield(heaters[i] - house)
                }
            }
            res = maxOf(res, distances.min())
        }
        
        return res
    }
}

File: 475.heaters/solution.go
func findRadius(houses []int, heaters []int) int {
    res := 0

    slices.Sort(heaters)

    for _, house := range houses {
        i, _ := slices.BinarySearch(heaters, house)

        distance := 0

        if i == 0 {
            distance = heaters[0] - house
        } else if i == len(heaters) {
            distance = house - heaters[len(heaters) - 1]
        } else {
            distance = max(distance, min(house - heaters[i - 1], heaters[i] - house))
        }

        res = max(res, distance)
    }

    return res
}